api/actions.ts

```diff
@@ -0,0 +1,327 @@
+import * as fs from 'fs';
+import {NTBBLadder} from './ladder';
+import {Replays} from './replays';
+import {toID} from './server';
+import * as tables from './tables';
+
+export const actions: {[k: string]: QueryHandler} = {
+	async register(params) {
+		const {username, password, cpassword, captcha} = params;
+		if (!username) {
+			throw new ActionError(`You must specify a username.`);
+		}
+		const userid = toID(username);
+		if (!/[a-z]/.test(userid)) {
+			throw new ActionError(`Your username must include at least one letter.`);
+		}
+		if (userid.startsWith('guest')) {
+			throw new ActionError(`Your username cannot start with 'guest'.`);
+		}
+		if (password.replace(/\s/ig, '').length < 5) {
+			throw new ActionError(`Your password must have at least 5 characters.`);
+		}
+		if (password !== cpassword) {
+			throw new ActionError(`Your passwords do not match.`);
+		}
+		if (toID(captcha) !== 'pikachu') {
+			throw new ActionError(`Answer the anti-spam question.`);
+		}
+		const regcount = await this.session.getRecentRegistrationCount(2 * 60 * 60);
+		if (regcount && regcount > 2) {
+			throw new ActionError(`You cannot register more than 2 names every 2 hours.`);
+		}
+		const user = await this.session.addUser(username, password);
+		if (user === null) {
+			throw new ActionError(`Your username is already taken.`);
+		}
+		const challengekeyid = parseInt(params.challengekeyid, 10) || -1;
+		const challenge = params.challstr || "";
+		if (!challenge) throw new ActionError(`Invalid challenge string argument.`);
+		const assertion = await this.session.getAssertion(userid, challengekeyid, user, challenge);
+		return {
+			assertion,
+			actionsuccess: !assertion.startsWith(';'),
+			curuser: await user.getData(),
+		};
+	},
+	async logout(params) {
+		if (
+			this.request.method !== "POST" || !params.userid ||
+			params.userid !== this.user.id || this.user.id === 'guest'
+		) {
+			return {actionsuccess: false};
+		}
+		await this.session.logout();
+		return {success: true};
+	},
+	async login(params) {
+		const challengeprefix = this.verifyCrossDomainRequest();
+		if (this.request.method !== 'POST') {
+			throw new ActionError(`For security reasons, logins must happen with POST data.`);
+		}
+		const userid = toID(params.name);
+		if (!userid || !params.pass) {
+			throw new ActionError(`incorrect login data, you need "name" and "pass" fields`);
+		}
+		const challengekeyid = parseInt(params.challengekeyid, 10) || -1;
+		const actionsuccess = await this.session.login(params.name, params.pass);
+		if (!actionsuccess) return {actionsuccess, assertion: false};
+		const challenge = params.challstr || "";
+		const assertion = await this.session.getAssertion(
+			userid, challengekeyid, null, challenge, challengeprefix
+		);
+		this.session.updateCookie();
+		return {
+			actionsuccess: true,
+			assertion,
+		};
+	},
+	async updateuserstats(params) {
+		const server = this.getServer(true);
+		if (!server) {
+			return {success: false};
```

Comment:
What's the difference between `success` and `actionsuccess`?

Reply:
Typo, really.